ledger/participant-integration-api/src/main/scala/platform/store/cache/StateCache.scala

```diff
@@ -0,0 +1,133 @@
+import com.daml.scalautil.Statement.discard
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future}
+
+/** This class is a wrapper around a Caffeine cache designed to handle
+  * correct resolution of concurrent updates for the same key.
+  */
+private[platform] case class StateCache[K, V](cache: Cache[K, V], metrics: Metrics)(implicit
+    ec: ExecutionContext
+) {
+  private val logger: ContextualizedLogger = ContextualizedLogger.get(getClass)
+  private[store] val pendingUpdates = mutable.Map.empty[K, PendingUpdates]
+
+  /** Fetch the corresponding value for an input key, if present.
+    *
+    * @param key the key to query for
+    * @return optionally [[V]]
+    */
+  def get(key: K)(implicit loggingContext: LoggingContext): Option[V] =
+    cache.getIfPresent(key) match {
+      case Some(value) =>
+        logger.debug(s"Cache hit for $key -> ${value.toString.take(100)}")
+        Some(value)
+      case None =>
+        logger.debug(s"Cache miss for $key ")
+        None
+    }
+
+  /** Update the cache synchronously.
+    *
+    * In face of multiple in-flight updates competing for the `key` (see [[putAsync()]]),
+    * this method updates the cache only if the to-be-inserted tuple is the most recent
+    * (i.e. it has `validAt` highest amongst the competing updates).
+    *
+    * @param key the key at which to update the cache
+    * @param validAt ordering discriminator for pending updates for the same key
+    * @param value the value to insert
+    */
+  def put(key: K, validAt: Long, value: V): Unit = pendingUpdates.synchronized {
+    val competingLatestForKey =
+      pendingUpdates
+        .get(key)
+        .map { pendingUpdate =>
+          val oldLatestValidAt = pendingUpdate.latestValidAt
+          pendingUpdate.latestValidAt =
+            if (pendingUpdate.latestValidAt < validAt) validAt else pendingUpdate.latestValidAt
+          oldLatestValidAt
+        }
+        .getOrElse(Long.MinValue)
+
+    if (competingLatestForKey < validAt) cache.put(key, value) else ()
+  }
+
+  /** Update the cache asynchronously.
+    *
+    * In face of multiple in-flight updates competing for the `key`,
+    * this method registers an async update to the cache
+    * only if the to-be-inserted tuple is the most recent
+    * (i.e. it has `validAt` highest amongst the competing updates).
+    *
+    * @param key the key at which to update the cache
+    * @param validAt ordering discriminator for pending updates for the same key
+    * @param eventualValue the eventual result
+    */
+  final def putAsync(key: K, validAt: Long, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Future[Unit] = pendingUpdates.synchronized {
+    val pendingUpdatesForKey = pendingUpdates.getOrElseUpdate(key, PendingUpdates.empty)
+    if (pendingUpdatesForKey.latestValidAt < validAt) {
+      pendingUpdatesForKey.latestValidAt = validAt
+      pendingUpdatesForKey.pendingCount += 1
```

Comment:
I don't understand why we need to have a list of pending updates per key, if only the latest one is ever inserted.

Reply:
I guess the choice of names is confusing. `PendingUpdates` tracks concurrent updates: it doesn't have a list, but it merely counts them and the most recent `validAt` of the bunch.

Reply:
You could use `PendingUpdatesState` as one option of a name that doesn't evoke the list connotation.

Reply:
I was going to go for `PendingUpdatesTracker` but `State` suffix is better